### ISSUE TITLE
New concept for entering recurrence rules / full-featured recurrence settings [$155]

### ISSUE THREAD

**Issue by georgehrke** [nextcloud/calendar#10]
The recurrence editor in the sidebar is currently being implemented.
It will follow the example of other calendar applications.

Although the current concept is pretty wide-spread, it's not really intuitive and we should create a concept for something more user-friendly.

This is definitely something to discuss at the next hackathon or conference.

**Comment by georgehrke** [comment]
![2016_09_22_12_58_57 copy](https://cloud.githubusercontent.com/assets/1250540/18958515/23cd2f22-8663-11e6-854e-ea10342bb640.png)

Mockups created at Nextcloud conf

**Comment by nickvergessen** [comment]
👍

**Comment by cyphar** [comment]
Looks like this is the nextcloud version of the request owncloud/calendar#770 (which I opened before I switched to Nextcloud).

**Comment by tcitworld** [comment]
Gnome Calendar just released a new UI for recurring events. https://www.youtube.com/watch?v=DoHLVmTvjFE

**Comment by georgehrke** [comment]
@tcitworld Thx a lot for the link!

I really like having "Monday - Friday". We should probably have "Saturday & Sunday" too :) 
The rest is pretty similar to our concept

**Comment by slideup-benni** [comment]
Why we won't support all of the RRULES defined in the ical spec?
http://www.kanzaki.com/docs/ical/rrule.html

For reference the google calendar interfaces:

![cal4](https://user-images.githubusercontent.com/12446028/29277432-ebdcc17c-8111-11e7-887a-0cacb1d12095.PNG)
![cal3](https://user-images.githubusercontent.com/12446028/29277434-ebe14e36-8111-11e7-9443-68c9abd12c67.PNG)
![cal2](https://user-images.githubusercontent.com/12446028/29277435-ebe36554-8111-11e7-9714-d2687a10e256.PNG)
![cal1](https://user-images.githubusercontent.com/12446028/29277433-ebddd3be-8111-11e7-851c-0ed5a085b70d.PNG)
![cal7](https://user-images.githubusercontent.com/12446028/29277437-ebf0d9fa-8111-11e7-9e27-7178fde9287a.PNG)
![cal6](https://user-images.githubusercontent.com/12446028/29277436-ebe9ec8a-8111-11e7-9f66-268bf5d99b2e.PNG)
![cal5](https://user-images.githubusercontent.com/12446028/29277438-ebf7b612-8111-11e7-907e-523912ecf2f6.PNG)

**Comment by georgehrke** [comment]
@slideup-benni Please check the mockup in the top. We will support all these rules.

I just want to add Mon-Fri and Sat&Sun to the quick drawer and not burry them in the advanced settings.

Btw, please don't post links to kanzaki iCal
It's still on RFC 2445 which is completely outdated.
We don't support RFC 2445

**Comment by georgehrke** [comment]
And I like the summary feature, we should have that too

**Comment by slideup-benni** [comment]
![Bountysource](https://api.bountysource.com/badge/issue?issue_id=39605191)

**Comment by ChrLau** [comment]
Would this also allow to create events that occur on "Every 3rd wednesday of the month" ?
That's what I would need. ;-)

**Comment by georgehrke** [comment]
Yes, it will allow to do that :)

**Comment by suntorytimed** [comment]
Will it be possible to add exceptions? I can already enter them in my GNOME calendar but Nextcloud and other calendars don't accept them as a valid value.

**Comment by georgehrke** [comment]
@suntorytimed see #7

**Comment by fazlerabbi37** [comment]
just shaw this issue and I am curious to know that would we be able to do "Repeat every n-time day/week/month/year" with this rule? 

![cal](https://user-images.githubusercontent.com/11360512/43886098-968766be-9bdc-11e8-874c-8198b9120c1f.png)

**Comment by MightyCreak** [comment]
@danielsteiner that's cool, but I think @georgehrke already has awaiting patches for that. Although that was a long time ago 😕

**Comment by georgehrke** [comment]
@danielsteiner That would be great! No one is working on this right now to my knowledge.

> but this looks a bit easier 

That's very optimistic. 😉
Recurrence-rules can get quite complex. You should take a look at [the recurrence rules in RFC5545](https://tools.ietf.org/html/rfc5545#section-3.3.10).
If you want to, we can also split it up so that I implement the manipulation of the actual calendar data and you implement the UI. Up to you.

One important thing though: The current calendar app (1.6.x) is running on Angular 1. This month, we will start migrating / rewriting the Calendar app in Vue.js. See #540.
I hope to have a first beta of Calendar 2.0, that will provide feature parity with 1.6 by end of October.
Would you be ok with waiting till end of October to start working on this feature?

**Comment by danielsteiner** [comment]
Absolutely. I'd actually prefer working with vue :D

**Comment by varac** [comment]
What's the state of this ? This is still a big blocker for me, since very basic recurrent events like `every 2. Wednesday of the month` is not possible. Please implement this soon !

**Comment by hedsteel** [comment]
I'm not a coder. Sorry I can't help with the project.

Would it be easier, faster to implement a quick half-fix that helps people create their own exceptions to recurring events? Would implementing an "end-date" feature for recurring events be easier than creating a whole bunch of options and rules for recurring events? Would implementing that right away solve a problem for users today while waiting on better features in the future?

Right now, if I have an event that happens ever Tuesday, but NOT on this particular Tuesday (thus the exception), if I try to delete just this particular Tuesday event, it deletes the entire series.
What if I could specify that there is an infinite recurring Tuesday event and I could simply edit when the event ends. 

Currently, I could edit the existing event to say that it ends after N occurrences, but to be honest, I have no idea how many occurrences of this event there are since it exists in the past indefinitely.

What if, instead of saying it ends after N occurrences, I could say that it ends LAST Tuesday (which this week would be Tuesday, 19 February 2019). The recurring event would not show up THIS Tuesday (tomorrow, 25 February). Then I could easily go to Tuesday of NEXT week (4 March) and recreate the infinitely repeating event going forward from that date.

That would allow me to have my exception to the recurring event.

**Comment by apiraino** [comment]
@hedsteel while what you describe could work, I'd rather see implemented the specification (I'm a coder, so my point of view is slightly different :-) I'm afraid that implementing exceptions could result in a completely different can of worms.

However, if I am not mistaken, a recurring event is just one line in a `iCal` record, example:
`RRULE:FREQ=WEEKLY;WKST=SU;UNTIL=YYYYMMDDTHHMMSSZ;BYDAY=MO,TU,WE,TH,FR`

While the feature is still to be implemented, I think I could workaround by creating a recurring event in some other calendar software that supports recurring events and then import the generated file into Nextcloud. It seems to work!  ¯\\_(ツ)_/¯

I've posted [on the forum](https://help.nextcloud.com/t/select-weekdays-in-every-week/45928/3?u=jman1) a template. Feedback on this is appreciated!

**Comment by Notsonoble** [comment]
This is definitely something I'd like to see improved. I've imported calendars with odd recurring events into nextcloud and it reads them fine, but not being able to edit them is frustrating.

At this point I'd even be okay with a "advanced" button that let me input a rrule manually. Would that be hard to implement?

**Comment by Sabanrab** [comment]
Is anyone actively working on this? I'm selfishly quite motivated to get this implemented, so if not, I wouldn't mind hitting it hard. Let me know!

**Comment by georgehrke** [comment]
Implemented in https://github.com/nextcloud/calendar/pull/926

**Comment by apiraino** [comment]
@georgehrke thanks :tada:

**Comment by tdoerschel** [comment]
@georgehrke thanks a lot for all the hard work on calendar 2.0!

**Comment by ChrLau** [comment]
Thanks also from me! So glad it's finally there. The 50 US-$ on Bountysource are worth it ;-)
